src/trio/_core/_tests/test_cancel.py

```diff
@@ -0,0 +1,151 @@
+# there's tons of cancellation testing in test_run, but that file is 3k lines long already...
+
+from math import inf
+
+import pytest
+
+import trio
+from trio import Cancelled
+from trio.lowlevel import current_task
+from trio.testing import RaisesGroup
+
+from .test_ki import ki_self
+
+
+async def test_cancel_reason() -> None:
+    with trio.CancelScope() as cs:
+        cs.cancel(reason="hello")
+        with pytest.raises(
+            Cancelled,
+            match=rf"^cancelled due to explicit with reason 'hello' from task {current_task()!r}$",
+        ):
+            await trio.lowlevel.checkpoint()
+
+    with trio.CancelScope(deadline=-inf) as cs:
+        with pytest.raises(Cancelled, match=r"^cancelled due to deadline"):
+            await trio.lowlevel.checkpoint()
+
+    with trio.CancelScope() as cs:
+        cs.deadline = -inf
+        with pytest.raises(
+            Cancelled,
+            match=r"^cancelled due to deadline",
+        ):
+            await trio.lowlevel.checkpoint()
+
+
+async def test_cancel_reason_nursery() -> None:
+    async def failing_task(task_status: trio.TaskStatus[trio.lowlevel.Task]) -> None:
+        task_status.started(current_task())
+        raise ValueError
+
+    async def cancelled_task(
+        fail_task: trio.lowlevel.Task, task_status: trio.TaskStatus
+    ) -> None:
+        task_status.started()
+        with pytest.raises(
+            Cancelled,
+            match=rf"^cancelled due to nursery with reason 'child task raised exception ValueError\(\)' from task {fail_task!r}$",
+        ):
+            await trio.sleep_forever()
+        raise TypeError
+
+    with RaisesGroup(ValueError, TypeError):
+        async with trio.open_nursery() as nursery:
+            fail_task = await nursery.start(failing_task)
+            await nursery.start(cancelled_task, fail_task)
+
+
+async def test_cancel_reason_nursery2() -> None:
```

Comment:
Could you add a comment explaining the difference? It took me a bit to see the difference is whether the task is cancelled before `trio.sleep_forever` or after.